File: 543.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int f(TreeNode* root){
        return root == NULL? 0 : (1 + max(f(root->left), f(root->right)));
    }

    void f1(TreeNode* root, int& sol){
        if(root  == NULL)return;
        sol = max(sol, 1 + f(root->left) + f(root->right));
        f1(root->left, sol);
        f1(root->right, sol);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        if(root == NULL) return 0;
        int sol = 0;
        f1(root, sol);
        return sol-1;
    }
};

File: Python/543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def f(self, root, sol):
        if root == None:
            return 0
        left = self.f(root.left, sol)
        right = self.f(root.right, sol)
        sol[0] = max(sol[0], 1 + left + right)
        return 1 + max(left, right)

    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        sol = []
        sol.append(0)
        self.f(root, sol)
        return sol[0]-1
